api/types/version_control_installer.go

```diff
@@ -0,0 +1,396 @@
+
+// IsLabelLiteral matches the subset of allowed label characters that never have a special
+// meaning.
+var IsLabelLiteral = regexp.MustCompile(`^[a-zA-Z0-9\.\-_\/:]+$`).MatchString
+
+func (e *ExecScript) Check() error {
+	if !IsStrictKebabCase(e.Type) {
+		// type name is used in a filename, so we need to be strict
+		// about its allowed characters.
+		return trace.BadParameter("invalid type %q in exec-script message", e.Type)
+	}
+
+	if e.Script == "" {
+		return trace.BadParameter("missing required field 'script' in exec-script message")
+	}
+
+	for name := range e.Env {
+		if !isValidEnvVarName(name) {
+			return trace.BadParameter("invalid env var name %q in exec-script message", name)
+		}
+	}
+
+	for _, name := range e.EnvPassthrough {
+		if !isValidEnvVarName(name) {
+			return trace.BadParameter("invalid env passthrough var name %q in exec-script message", name)
+		}
+	}
+
+	return nil
+}
+
+// VersionControlInstallerKind is the kind of a version control installer.
+type VersionControlInstallerKind string
+
+const (
+	// InstallerKindNone indicates that kind was not specified. The meaning of this is
+	// context-dependent (e.g. a filter treats InstallerKindNone as 'match all').
+	InstallerKindNone VersionControlInstallerKind = ""
+
+	// InstallerKind is the local script installer kind.
+	InstallerKindLocalScript VersionControlInstallerKind = "local-script"
+)
+
+func (t VersionControlInstallerKind) Check() error {
+	switch t {
+	case InstallerKindNone:
+	case InstallerKindLocalScript:
+	default:
+		return trace.BadParameter("unknown version control installer kind %q", t)
+	}
+	return nil
+}
+
+// Iter iterates all installers in the set.
+func (s *VersionControlInstallerSet) Iter(fn func(i VersionControlInstaller)) {
+	for _, installer := range s.LocalScript {
+		fn(installer)
+	}
+}
+
+// VersionControlInstallerStatus indicates the current status of an installer in terms of
+// its theoretical runnability.
+type VersionControlInstallerStatus string
+
+const (
+	// InstallerStatusNone indicates that a status was not specified. The meaning of this
+	// is context-dependent.
+	InstallerStatusNone VersionControlInstallerStatus = ""
+
+	// InstallerStatusOk indicates that the installer can be used.
+	InstallerStatusOk VersionControlInstallerStatus = "ok"
+
+	// InstallerStatusLocked indicates that the installer should not be used.
+	InstallerStatusLocked VersionControlInstallerStatus = "locked"
+)
+
+func (s VersionControlInstallerStatus) Check() error {
+	// allow unkwnown statuses, but keep allowed characters in-line with
+	// the naming conventions of known statuses.
+	if !IsStrictKebabCase(string(s)) {
+		return trace.BadParameter("invalid version control installer status %q", s)
+	}
+
+	return nil
+}
+
+func (s VersionControlInstallerStatus) IsKnownStatus() bool {
+	switch s {
+	case InstallerStatusNone:
+	case InstallerStatusOk:
+	case InstallerStatusLocked:
+	default:
+		return false
+	}
+
+	return true
```

Comment:
nit: I'd find it easier to read if the return true wasn't so far away from the valid statuses.
```suggestion
	switch s {
	case InstallerStatusNone, InstallerStatusOk, InstallerStatusLocked:
		return true
	default:
		return false
	}
```